File: aoc-py/solutions/day5.py
"""
Day 5: Cafeteria

https://adventofcode.com/2025/day/5
"""
__all__ = ('Day5',)

from typing import ClassVar

from ..solution import Solution

class Day5(Solution):
    NAME: ClassVar[str] = 'Cafeteria'

    def part_one(self, inp: str) -> int:
        ranges, ingredients = inp.split('\n\n')
        ranges = [
            tuple(int(p) for p in r.split('-'))
            for r in ranges.splitlines()
        ]

        return sum(
            any(a <= int(ingredient) <= b for a, b in ranges)
            for ingredient in ingredients.splitlines()
        )

    def part_two(self, inp: str) -> int:
        # sort ranges so overlapping ranges will be found consecutively
        ranges = sorted(
            tuple(int(p) for p in r.split('-'))
            for r in inp.split('\n\n')[0].splitlines()
        )

        i = 1
        # merge overlapping ranges together (union)
        while i < len(ranges):
            a0, b0 = ranges[i - 1]
            a, b = ranges[i]

            # intersection of [a0, b0] and [a, b]:
            # a0--a--b--b0
            # a0--a--b0--b
            # a0--b0--a--b
            if a0 <= a <= b0:
                # union [a0, b0] and [a, b] into ranges[i - 1]
                ranges[i - 1] = (a0, max(b, b0))
                del ranges[i]
            else:
                # only traverse forwards if current node not deleted
                # otherwise, no need, since next el will get swapped back automatically during `del ranges[i]`
                i += 1

        # since all ranges are non-overlapping
        # we can just sum their numeric ranges
        return sum(b - a + 1 for a, b in ranges)

    def run(self, inp: str) -> None:
        print('Part 1:', p1 := self.part_one(inp))
        print('Part 2:', p2 := self.part_two(inp))

        assert p1 == 874
        assert p2 == 348548952146313

File: src/bin/day5.rs
//! Day 5: Cafeteria
//!
//! <https://adventofcode.com/2025/day/5>
use aoc_2025::Solution;

pub struct Day5;

impl Solution for Day5 {
    const NAME: &'static str = "Cafeteria";

    fn part_one(&self, inp: &str) -> Self::OutputP1 {
        let inp = inp.replace('\r', "");

        let (ranges, ingredients) = inp
            .split_once("\n\n")
            .unwrap();

        let ranges = ranges
            .lines()
            .map(|line| {
                let (a, b) = line.split_once('-').unwrap();

                (a.parse().unwrap(), b.parse().unwrap())
            })
            .collect::<Vec<(usize, usize)>>();

        ingredients
            .lines()
            .filter(|ingredient| ranges
                .iter()
                .any(|&(a, b)| {
                    let ingredient = ingredient.parse().unwrap();
                    a <= ingredient && ingredient <= b
                })
            )
            .count()
    }

    fn part_two(&self, inp: &str) -> Self::OutputP2 {
        let inp = inp.replace('\r', "");

        let mut ranges = inp
            .split_once("\n\n")
            .unwrap()
            .0
            .lines()
            .map(|line| {
                let (a, b) = line.split_once('-').unwrap();

                (a.parse().unwrap(), b.parse().unwrap())
            })
            .collect::<Vec<(usize, usize)>>();

        ranges.sort_unstable();

        let mut i = 1;
        while i < ranges.len() {
            let (a0, b0) = ranges[i - 1];
            let (a, b) = ranges[i];

            if a0 <= a && a <= b0 {
                ranges[i - 1] = (a0, b.max(b0));
                ranges.remove(i);
            } else {
                i += 1;
            }
        }

        ranges
            .into_iter()
            .map(|(a, b)| b - a + 1)
            .sum()
    }

    fn run(&self, inp: String) {
        let p1 = self.part_one(&inp);
        let p2 = self.part_two(&inp);

        println!("Part 1: {p1}");
        println!("Part 2: {p2}");

        assert_eq!(p1, 874);
        assert_eq!(p2, 348_548_952_146_313);
    }
}

fn main() {
    aoc_2025::run_day(5, &Day5);
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test() { main(); }
}